keepalived: update to 1.2.19, add procd init format, and UCI config parsing

commodo opened the pull request:
@psycho-nico please review :)
@sbyx @thess [or other maintainers] if this needs to be squashed into a single commit, I'm fine with that. I'll do it after the maintainer of this package gives blessing.

Signed-off-by: Alexandru Ardelean <ardeleanalex@gmail.com>

champtar commented:
Hi @commodo,
is it still possible to use old style config file or is it uci only after your patch?

commodo commented:
Hmm, I'd need to update the patch to allow the use of the previous config style.
Because as it is now with this patch, the entire `/etc/keepalived/keepalived.conf` file will be overwritten on start/restart/reload.
We use UCI (and try to use it as much as possible)  in our system, because writing code that generates UCI style configs is very simple (and easy to maintain).

I'll think about how to do this properly.
Maybe with an env var, or something in `/etc/rc.local`, or some other config file ?
Suggestions are welcome.

However, whatever the implementation will be, one question remains:
What should the default be now : UCI or old config style ?

I won't add a vote, since I can adapt to whatever the maintainer and/or majority wants.

Thanks
Alex

sbyx commented:
Please do not dynamically create config files from uci on persistent memory (/etc), this will just unnecessarily create flash writes at every startup. If you dynamically create config files from uci, those should be written to the ramdisk (/tmp).

champtar commented:
I know openvpn support both, you just tell in the uci conf file the old style conf. Maybe if there is the old file, take it, else take the uci one, or vice versa. Also keep the conffiles part, else you risk loosing the old conf file.

commodo commented:
@sbyx will adjust to write config to /tmp ; our internal setup differs from core so I was not aware of this
@champtar will take a look at openvpn ; maybe I'll use it if I can't come up with a more comfortable implementation

thanks
alex

commodo commented:
re-spinned;

so, logic as is now:  
* /tmp/keepalived.conf will be used by keepalived
* if /etc/keepalived.conf exists, it will be copied over /tmp/keepalived.conf
* otherwise /tmp/keepalived.conf gets generated via init.d script from /etc/config/network
* if no /etc/config/network exists, /tmp/keepalived.conf will be empty (just the first line gets added)

tested it now ;
but will run some more tests tomorrow, since I'll have to integrate it in our system

@champtar is this approach ok for you ?
@sbyx : using /tmp
@psycho-nico will wait for your review/go-ahead/nack

Thanks
Alex

commodo commented:
re-re-spinned; 
re-tested and all seems to be fine

added option to disable the `keepalived.sample` config from being installed
we don't need it, but maybe it's good to have it 

in any case, I'm open to suggestions on it; 
if anyone wants me to remove it, please say so

champtar commented on net/keepalived/files/keepalived.init:
```diff
@@ -2,13 +2,355 @@
+	procd_append_param command -f "$KEEPALIVED_CONF"
+
+	if [ -f /etc/keepalived.conf ]; then
+		cp -f /etc/keepalived.conf $KEEPALIVED_CONF
```
you could use "ln -s"

champtar replied:
it's /etc/keepalived/keepalived.conf no?

commodo replied:
`ln -s` sounds better

Why not move it to  `/etc/keepalived.conf` ?
keepalived does not use more than 1 conf file anyway ? 

I have no strong opinions about keeping it in `/etc/keepalived/keepalived.conf` or moving to `/etc/keepalived.conf`.
It's just a proposal.

champtar replied:
i'm ok to move it but:
it must be transparent to the user (post install hook or ...)
and
you must keep it on upgrades (conffiles sections)

champtar commented on net/keepalived/Makefile:
```diff
@@ -32,12 +32,12 @@ define Package/keepalived
 endef
 
-define Package/keepalived/conffiles
-/etc/keepalived/keepalived.conf
```
you must keep this conffiles section else next time you update you will loose this conf file :(

commodo replied:
well, then let's agree here on a default:
do we keep keepalived's keepalived.conf or 
use the new UCI config as default ?

champtar replied:
we keep both, you don't want to lose user conf ...

champtar commented:
hi @commodo,
your sample file is less than 1Ko after gzip, current keepalived ipk is 67Ko, i would just ship the sample commented out as /etc/config/keepalived
see also https://dev.openwrt.org/browser/trunk/package/network/services/openvpn/files/openvpn.config

commodo commented:
Sounds good to me to ship with stuff commented out.

commodo commented:
@champtar 
re-spinned again
this time, added `option uci_enabled`  in the `global_defs` section

tested with `uci_enabled` 1 and 0 and option removed;
option not present and value 0 means that `ln -s /etc/keepalived/keepalived.conf $KEEPALIVED_CONF`  is used

since backwards compat is a consideration, I've reverted to just `/etc/keepalived/keepalived.conf`
we'll probably use `/etc/config/keepalived` anyway, so it does not matter much for us

commodo commented:
@psycho-nico ping :)

psycho-nico commented:
@commodo
Please keep the commit updating the package separate from the others.
I vote for a single uci config file with all/most stuff commented out.
I'm not convinced by the "uci_enabled" UCI option... Why not use smth like "alt_config_file" or "legacy_config_file" pointing to "/etc/keepalived/keepalived.conf". When set, it would be used as $KEEPALIVED_CONF and other UCI options would be ignored.

commodo commented:
@psycho-nico 
I just saw your comments ; I think I was on vacation at the time.
I'll close this PR , spin another 2, first to update the UCI config file format.
And then the second to do the update, in order to keep the PKG_RELEASE to 1

The `alt_config_file` option sounds good to me.
`legacy_config_file` sounds partially incorrect (semantically speaking) because it's technically not legacy so much as a UCI wrapper added in-between.
